open-craft/opencraft
# Add secret key generation to OpenCraft IM

This PR adds logic to automatically generate secret keys for a variety of services populated by OpenCraft IM.

**Dependencies**: None

**Testing instructions**:
1. Using a copy of OCIM as it exists right now, set up an instance and spawn an appserver. There's no need to wait for the appserver to provision; just look at the appserver's vars and observe that items like `EDXAPP_EDXAPP_SECRET_KEY` are not present.
2. Pull this branch into your existing OCIM instance and migrate it.
3. Spawn a new appserver on the existing instance. Again, no need to wait for it to provision; just look at the vars in OCIM and observe that items like `EDXAPP_EDXAPP_SECRET_KEY` are still missing - for instances that don't have a master secret key set, we don't want to change things.
4. Create a new instance and spawn an appserver. Observe that variables like `EDXAPP_EDXAPP_SECRET_KEY` and `FORUM_API_KEY` are populated with seemingly-random values.
5. Spawn a new appserver on the same instance; observe that its secret key vars have the same values as the first one.

**Author notes and concerns**:
1. The current setup is a little awkward. Because the secret keys are derived based on the variable names, we need to have those names in source, not just in a template. I explored other options, but the current setup is robust and allowed me to get the code deployed. I am very open to suggestions on how to improve it - one thought I had was by feeding in the secret keys we're interested in filling as environment variables and building the YAML string entirely in Python, but that seems a bit obfuscated.

**Reviewers**
